language/move-stdlib/src/natives/debug.rs

```diff
@@ -60,6 +61,16 @@ fn native_print(
     #[cfg(feature = "testing")]
     {
         let ty = ty_args.pop().unwrap();
+        
```

Comment:
There seem to be a lot of compilation errors:

```
error[E0425]: cannot find value `_context` in this scope
  --> language/move-stdlib/src/natives/debug.rs:75:30
   |
75 |         match is_string_type(_context, &ty, move_std_addr) {
   |                              ^^^^^^^^ help: a local variable with a similar name exists: `context`

error[E0308]: mismatched types
   --> language/move-stdlib/src/natives/debug.rs:69:107
    |
69  |         let move_struct = MoveStruct::simple_deserialize(&args[0].simple_serialize(&old_layout).unwrap(), &old_layout).unwrap();
    |                           ------------------------------ arguments to this function are incorrect         ^^^^^^^^^^^ expected enum `MoveStructLayout`, found enum `MoveTypeLayout`
    |
    = note: expected reference `&MoveStructLayout`
               found reference `&MoveTypeLayout`
note: associated function defined here
   --> /Users/alinush/repos/move/language/move-core/types/src/value.rs:164:12
    |
164 |     pub fn simple_deserialize(blob: &[u8], ty: &MoveStructLayout) -> AResult<Self> {
    |            ^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> language/move-stdlib/src/natives/debug.rs:72:53
    |
72  |         let decorated_struct = move_struct.decorate(new_layout);
    |                                            -------- ^^^^^^^^^^ expected `&MoveStructLayout`, found enum `MoveTypeLayout`
    |                                            |
    |                                            arguments to this function are incorrect
    |
note: associated function defined here
   --> /Users/alinush/repos/move/language/move-core/types/src/value.rs:168:12
    |
168 |     pub fn decorate(self, layout: &MoveStructLayout) -> Self {
    |            ^^^^^^^^
```

Reply:
Tried fixing it with this diff. It now compiles, but...

```
diff --git a/language/move-stdlib/nursery/sources/debug.move b/language/move-stdlib/nursery/sources/debug.move
index 3acb2c856..0c190748c 100644
--- a/language/move-stdlib/nursery/sources/debug.move
+++ b/language/move-stdlib/nursery/sources/debug.move
@@ -16,4 +16,22 @@ module std::debug {
         let str = string::utf8(str_bytes);
         print<string::String>(&str);
     }
+
+    #[test_only]
+    struct TestStruct has drop {
+        number: u8,
+        bytes: vector<u8>,
+        name: string::String,
+    }
+
+    #[test]
+    fun test_print_annotated_struct() {
+        let obj = TestStruct {
+            number: 255u8,
+            bytes: x"c0ffee",
+            name: string::utf8(b"Hello, sane Move debugging!"),
+        };
+
+        print(&obj);
+    }
 }
diff --git a/language/move-stdlib/src/natives/debug.rs b/language/move-stdlib/src/natives/debug.rs
index f5d938348..417c3dac6 100644
--- a/language/move-stdlib/src/natives/debug.rs
+++ b/language/move-stdlib/src/natives/debug.rs
@@ -4,7 +4,7 @@
 
 use crate::natives::helpers::make_module_natives;
 use move_binary_format::errors::PartialVMResult;
-use move_core_types::value::MoveStruct;
+use move_core_types::value::{MoveStruct, MoveTypeLayout};
 use move_core_types::{
     account_address::AccountAddress, gas_algebra::InternalGas, language_storage::TypeTag,
 };
@@ -64,36 +64,53 @@ fn native_print(
         
         // if ty is a struct type:
         // get type layout in VM format (raw BCS, no field names) 
-        let old_layout = context.type_to_type_layout(&ty)?.unwrap();
-        // get BCS bytes for the struct
-        let move_struct = MoveStruct::simple_deserialize(&args[0].simple_serialize(&old_layout).unwrap(), &old_layout).unwrap();
-        // get type layout in decorated format (with field names)
-        let new_layout = context.type_to_fully_annotated_layout(&ty)?.unwrap();
-        let decorated_struct = move_struct.decorate(new_layout);
-        println!("[debug] {}", decorated_struct);
-
-        match is_string_type(_context, &ty, move_std_addr) {
-            true => {
-                let r = pop_arg!(args, StructRef);
-
-                let field_ref = r
-                    .borrow_field(0)? // borrows the 'bytes' field of the std::string::String struct
-                    .value_as::<VectorRef>()?;
-
-                let bytes = field_ref.as_bytes_ref();
-
-                // This is safe because we guarantee the bytes to be utf8.
-                let str = unsafe { std::str::from_utf8_unchecked(bytes.as_slice()) };
-
-                println!("[debug] {}", str);
-            }
-            false => {
-                let r = pop_arg!(args, Reference);
-                let mut buf = String::new();
-                print_reference(&mut buf, &r)?;
-                println!("[debug] {}", buf);
+        let old_type_layout = context.type_to_type_layout(&ty)?.unwrap();
+
+        println!("{}", &args[0]);
+        println!("{}", old_type_layout);
+
+        match &old_type_layout {
+            MoveTypeLayout::Struct(old_struct_layout) => {
+                // get BCS bytes for the struct
+                let blob = &args[0].simple_serialize(&old_type_layout).unwrap();
+                let move_struct = MoveStruct::simple_deserialize(blob, old_struct_layout).unwrap();
+
+                // get type layout in decorated format (with field names)
+                let new_type_layout = context.type_to_fully_annotated_layout(&ty)?.unwrap();
+                let new_struct_layout = match &new_type_layout {
+                    MoveTypeLayout::Struct(struct_layout) => struct_layout,
+                    _ => panic!("ooppsie; this shouldn't happen.")
+                };
+
+                let decorated_struct = move_struct.decorate(new_struct_layout);
+                println!("[debug] {}", decorated_struct);
+            },
+            _ => {
+                match is_string_type(context, &ty, move_std_addr) {
+                    true => {
+                        let r = pop_arg!(args, StructRef);
+
+                        let field_ref = r
+                            .borrow_field(0)? // borrows the 'bytes' field of the std::string::String struct
+                            .value_as::<VectorRef>()?;
+
+                        let bytes = field_ref.as_bytes_ref();
+
+                        // This is safe because we guarantee the bytes to be utf8.
+                        let str = unsafe { std::str::from_utf8_unchecked(bytes.as_slice()) };
+
+                        println!("[debug] {}", str);
+                    }
+                    false => {
+                        let r = pop_arg!(args, Reference);
+                        let mut buf = String::new();
+                        print_reference(&mut buf, &r)?;
+                        println!("[debug] {}", buf);
+                    }
+                }
             }
-        }
+        };
+
     }
 
     Ok(NativeResult::ok(gas_params.base_cost, smallvec![]))

```

...but the `&args[0].simple_serialize(&old_type_layout).unwrap()` call panics when unwrapping a `None`:

```
(2, [U8(255), [192, 255, 238], [[72, 101, 108, 108, 111, 44, 32, 115, 97, 110, 101, 32, 77, 111, 118, 101, 32, 100, 101, 98, 117, 103, 103, 105, 110, 103, 33]]])
{ 0: u8, 1: vector<u8>, 2: { 0: vector<u8>, }, }
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', language/move-stdlib/src/natives/debug.rs:75:72
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

You can reproduce with a `cargo test --  --nocapture` in `language/move-stdlib/nursery/sources`